### PR TITLE
Best Practices with Git

### PR DESCRIPTION
**Pull request by darrenwee**
This is the first draft of my book chapter on git best practices. I've highlighted some important pointers that I've learnt about (most of them the hard way) since I started using git.

Can you take a look @vivekscl :grinning: 

### [view the rendered version here](https://github.com/darrenwee/learningresources/blob/git-best-practices/contents/revisionControl/bestPracticesGit.md)

### CODE REVIEW

File: contents/revisionControl/bestPracticesGit.md
Comment (vivekscl) [inline contents/revisionControl/bestPracticesGit.md]: ... reasoning behind **the** changes made...

File: contents/revisionControl/bestPracticesGit.md
Comment (vivekscl) [inline contents/revisionControl/bestPracticesGit.md]: Maybe you could explicitly say to type the following in a terminal like `do one of either in your terminal`?

File: contents/revisionControl/bestPracticesGit.md
Comment (vivekscl) [inline contents/revisionControl/bestPracticesGit.md]: You stated `bash` here but there doesn't seem to be any syntax highlighting, so is this needed? Furthermore, the `bash` keyword wasn't used in the above section about `Set Up Your Editor for Commit Messages`.

File: contents/revisionControl/bestPracticesGit.md
Comment (vivekscl) [inline contents/revisionControl/bestPracticesGit.md]: Not sure what a hanging indent is though :sweat_smile: . Maybe you should explain it?

File: contents/revisionControl/bestPracticesGit.md
Comment (vivekscl) [inline contents/revisionControl/bestPracticesGit.md:166]: Maybe add this as a note instead?

File: contents/revisionControl/bestPracticesGit.md
Comment (vivekscl) [inline contents/revisionControl/bestPracticesGit.md]: Does `with little to no side effects` sound better?

File: contents/revisionControl/bestPracticesGit.md
Comment (vivekscl) [inline contents/revisionControl/bestPracticesGit.md]: Should it be, `collaborate with others easily` instead? Since you're emphasizing the benefits, just stating `collaborate with others` doesn't seem as beneficial.

File: contents/revisionControl/bestPracticesGit.md
Comment (vivekscl) [inline contents/revisionControl/bestPracticesGit.md]: Maybe you could say `Always avoid rewriting the public history` or `Always avoid rewriting the published history` instead? Putting both terms in there seems redundant.

Comment (vivekscl) [inline contents/revisionControl/bestPracticesGit.md]: If there is an `Exceptions` section below, is this part needed? 
`unless you have a very good reason, you are very sure of what you are doing, or you will not affect anyone else`

File: contents/revisionControl/bestPracticesGit.md
Comment (vivekscl) [inline contents/revisionControl/bestPracticesGit.md]: To standardize the headers, `when ...` should have the `w` capitalized and the space removed between `when` and `...` like so:
**W**hen**...**

File: contents/revisionControl/bestPracticesGit.md
Comment (vivekscl) [inline contents/revisionControl/bestPracticesGit.md]: An unnecessary period at the end of this line.

File: contents/revisionControl/bestPracticesGit.md
Comment (vivekscl) [reply contents/revisionControl/bestPracticesGit.md]: Since point 2 has a `:` at the end, point 1 should have a `:` at the end.

File: contents/revisionControl/bestPracticesGit.md
Comment (vivekscl) [inline contents/revisionControl/bestPracticesGit.md]: Missing `:` at the end

File: contents/revisionControl/bestPracticesGit.md
Comment (vivekscl) [inline contents/revisionControl/bestPracticesGit.md]: Missing `:` at the end

File: contents/revisionControl/bestPracticesGit.md
Comment (vivekscl) [inline contents/revisionControl/bestPracticesGit.md]: Maybe there should be a newline after the `:` and bullet point like the other sections?
deepinscreenshot_select-area_20180305162937

File: contents/revisionControl/bestPracticesGit.md
Comment (vivekscl) [inline contents/revisionControl/bestPracticesGit.md]: Maybe what stash does should be explicitly stated?
you can use the `git stash` command **to save your code** instead of...

File: contents/revisionControl/bestPracticesGit.md
Comment (vivekscl) [inline contents/revisionControl/bestPracticesGit.md]: `P`rogramming practices

**Review by vivekscl — CHANGES_REQUESTED** [review]
Several standardization issues and suggestions. 
Content-wise, it's really informative and well structured :+1:

File: contents/revisionControl/bestPracticesGit.md
Comment (Zhiyuan-Amos) [inline contents/revisionControl/bestPracticesGit.md]: You can use `git bisect` even without rebasing to master right? Cos you can git bisect on `master`, and once you find out which `merge` is buggy, then `git bisect` the `merge/PR`.

File: contents/revisionControl/bestPracticesGit.md
Comment (Zhiyuan-Amos) [inline contents/revisionControl/bestPracticesGit.md]: This is already mentioned above

> It is generally considered good practice to rebase your feature branch onto whatever branch you're trying to patch _before_ you make the pull request, **resolving any conflicts that arise.**

File: contents/revisionControl/bestPracticesGit.md
Comment (Zhiyuan-Amos) [inline contents/revisionControl/bestPracticesGit.md]: The title can be better phrased because this seems more of like `How to add a remote` rather than `use remotes effectively`.

File: contents/revisionControl/bestPracticesGit.md
Comment (Zhiyuan-Amos) [inline contents/revisionControl/bestPracticesGit.md:228]: This segment seems a bit short compared to the rest, and also the idea of `force pushing` to rewrite master branch aren't mentioned in the links as well.

File: contents/revisionControl/bestPracticesGit.md
Comment (Zhiyuan-Amos) [inline contents/revisionControl/bestPracticesGit.md]: You can give examples of good commits. Perhaps... Get from `AddressBook`? :P

File: contents/revisionControl/bestPracticesGit.md
Comment (Zhiyuan-Amos) [inline contents/revisionControl/bestPracticesGit.md]: You can mention why we should set the editor to wrap after 72 characters.

File: contents/revisionControl/bestPracticesGit.md
Comment (tshradheya) [inline contents/revisionControl/bestPracticesGit.md]: One good thing I have read in many articles related to writing git commit messages is to make sure the message makes sense when read as follows:

>If applied, this commit will `your commit message here`

Maybe can add this here?

File: contents/revisionControl/bestPracticesGit.md
Comment (darrenwee) [reply contents/revisionControl/bestPracticesGit.md]: I will not implement this change as it's the original title of the link, so I want to preserve it as-is.

File: contents/revisionControl/bestPracticesGit.md
Comment (darrenwee) [reply contents/revisionControl/bestPracticesGit.md]: Haha sure, thanks!!

File: contents/revisionControl/bestPracticesGit.md
Comment (darrenwee) [reply contents/revisionControl/bestPracticesGit.md]: This point (line 227) is supposed to be about the benefit of rebasing onto the base branch _before_ making a PR. Resolving conflicts during the rebase process guarantees that your branch is compatible with the base branch that you're patching, so it's not quite the same point I'm trying to put across. One is cause and the other is effect. What do you think?

Comment (Zhiyuan-Amos) [reply contents/revisionControl/bestPracticesGit.md]: Ah ok I see! :)

File: contents/revisionControl/bestPracticesGit.md
Comment (darrenwee) [reply contents/revisionControl/bestPracticesGit.md:166]: I left it as-is because I'm presenting it as an alternative to committing WIP code, so it should be in the body.

File: contents/revisionControl/bestPracticesGit.md
Comment (darrenwee) [reply contents/revisionControl/bestPracticesGit.md]: I added it outside of the commit message code block so that it stands out more.

File: contents/revisionControl/bestPracticesGit.md
Comment (darrenwee) [reply contents/revisionControl/bestPracticesGit.md:228]: I beefed this section up and added a bit on force pushing.

File: contents/revisionControl/bestPracticesGit.md
Comment (darrenwee) [reply contents/revisionControl/bestPracticesGit.md]: The syntax highlighting is for the bash comments so that it doesn't look the same as the git commands.

**Comment by darrenwee** [comment]
@damithc I'm ready for a review!

File: contents/revisionControl/bestPracticesGit.md
Comment (Zhiyuan-Amos) [inline contents/revisionControl/bestPracticesGit.md]: the comma here is kinda weird.

**Review by Zhiyuan-Amos — COMMENTED** [review]
Sorry, don't have time to give a thorough look, so I won't give an explicit approve / request changes :<

File: contents/revisionControl/bestPracticesGit.md
Comment (darrenwee) [reply contents/revisionControl/bestPracticesGit.md]: I will revise the structure to clean this part up and remove the redundant `while` and commas.

**Comment by damithc** [comment]
Had a quick look. No major objections. Get approval from a peer first?

**Review by vivekscl — APPROVED** [review]
Looks good to me!

**Comment by darrenwee** [comment]
@damithc Sorry about that, it's approved now :smile:

File: contents/revisionControl/bestPracticesGit.md
Comment (damithc) [inline contents/revisionControl/bestPracticesGit.md]: One Logical Change Per Commit?

File: contents/revisionControl/bestPracticesGit.md
Comment (damithc) [inline contents/revisionControl/bestPracticesGit.md]: Fixing multiple bugs in one commit contradicts 'one logical change per commit'?

File: contents/revisionControl/bestPracticesGit.md
Comment (damithc) [inline contents/revisionControl/bestPracticesGit.md]: `your commit message summary here`? 

Also, can give an example

File: contents/revisionControl/bestPracticesGit.md
Comment (damithc) [inline contents/revisionControl/bestPracticesGit.md]: This may be useful https://oss-generic.github.io/process/docs/FormatsAndConventions.html#commit-message

File: contents/revisionControl/bestPracticesGit.md
Comment (damithc) [inline contents/revisionControl/bestPracticesGit.md]: `Hide the Sausage` or `Hide the Sausage Making`?

File: contents/revisionControl/bestPracticesGit.md
Comment (damithc) [inline contents/revisionControl/bestPracticesGit.md]: It is good to give an overview of each resource like you have done here. Try to do it more. But may be not use italics?

**Review by damithc — CHANGES_REQUESTED** [review]
Some suggestions

File: contents/revisionControl/bestPracticesGit.md
Comment (darrenwee) [reply contents/revisionControl/bestPracticesGit.md]: I decided to scrap the old format for describing a good commit message and I took the list from that link and reorganized it a bit.

File: contents/revisionControl/bestPracticesGit.md
Comment (darrenwee) [reply contents/revisionControl/bestPracticesGit.md]: I wasn't sure what was the best way to do this, so I took out the italics and uses a hyphen to delimit the start of the overview. I only give an overview for the more important/interesting resources, though. Some like the StackOverflow threads are self-explanatory.

**Comment by darrenwee** [comment]
@damithc I made some structural changes to the section on commit messages; otherwise everything else are minor edits as per your comments :smile:

**Review by damithc — CHANGES_REQUESTED** [review]
Looks good. As you have given very detailed guidelines for  commit message writing, it would be good to give one or two concrete examples (some people prefer to learn from examples). You can reuse the examples from oss-generic page (remember to cite the source).

File: contents/revisionControl/bestPracticesGit.md
Comment (rachx) [inline contents/revisionControl/bestPracticesGit.md]: Link broken

**Comment by darrenwee** [comment]
I've fixed the broken reference and added some examples of good commit messages! I was tempted to add some more cool git tools like `git bisect` or the `reflog` but I think that can be left for a different chapter (or someone else to write).

**Review by damithc — APPROVED** [review]
LGTM

**Pull request merged**
